src/components/[guild]/JoinModal/hooks/useOauthPopupWindow.ts

```diff
@@ -48,129 +18,116 @@ type OAuthOptions = {
-        try {
-          const {
-            data,
-            type,
-            csrfToken: receivedCsrfToken,
-          } = JSON.parse(window.localStorage.getItem(dataKey))
-
-          addDatadogAction(
-            `CSRF - Main window received CSRF token: ${receivedCsrfToken}. Should equal: ${csrfToken}`
-          )
-
-          if (type === "OAUTH_ERROR") {
-            clearInterval(interval)
-            const title = data?.error ?? "Unknown error"
-            const errorDescription = data?.errorDescription ?? ""
-            reject({ error: title, errorDescription })
-            toast({ status: "error", title, description: errorDescription })
-          }
-          if (type === "OAUTH_SUCCESS") {
-            clearInterval(interval)
-
-            if (receivedCsrfToken !== csrfToken) {
-              const title = "Error"
-              const errorDescription =
-                "Authentication attempt can't be validated. Refresh and try connecting again"
-
-              addDatadogError(`OAuth error - ${title}`, {
-                error: errorDescription,
-                receivedCsrfToken,
-                csrfToken,
-              })
-              reject({ error: title, errorDescription })
-              toast({ status: "error", title, description: errorDescription })
-            } else {
-              resolve({
-                redirect_url: redirectUri,
-                scope: oauthOptions.scope,
-                ...(data as OAuthResponse),
-              })
-            }
-          }
-        } catch {}
-      }, 500)
-    })
-      .then(setAuthData)
-      .catch(setError)
-      .finally(() => {
-        if (windowInstanceOpenInitially) {
-          const closeInterval = setInterval(() => {
-            if (windowInstance.closed) {
-              setIsAuthenticating(false)
-              clearInterval(closeInterval)
-            }
-          }, 500)
+    window.localStorage.setItem(
+      localStorageKey,
+      JSON.stringify(infoToPassInLocalStorage)
+    )
+
+    const channel = new BroadcastChannel(csrfToken)
+
+    const hasReceivedResponse = new Promise<void>((resolve) => {
+      channel.onmessage = (event: MessageEvent<Message>) => {
+        const { type, data } = event.data
+
+        if (type === "OAUTH_ERROR") {
+          setOauthState({
+            isAuthenticating: false,
+            error: data,
+            authData: null,
+          })
+        } else {
+          setOauthState({
+            isAuthenticating: false,
+            error: null,
+            authData: data,
+          })
         }
 
-        window.localStorage.removeItem(dataKey)
-        setIsAuthenticating(false)
-        window.localStorage.setItem(shouldCloseKey, "true")
-        mutateCSRFToken(randomBytes(16).toString("hex"))
-      })
-  }, [windowInstance])
+        resolve()
+      }
+    })
+
+    onOpen(
+      `${url}?${Object.entries(oauthOptions)
+        .map(([key, value]) => `${key}=${value}`)
```

Comment:
I think we could use URLSearchParams here to make it a bit cleaner, maybe something like this?

```ts
const searchParams = new URLSearchParams({
  ...oauthOptions,
  redirect_uri: redirectUri,
  state: csrfToken
}).toString()

onOpen(`${url}?${searchParams}`)
```

Reply:
Thank you, this is indeed much cleaner!

Reply:
Made this change in https://github.com/agoraxyz/guild.xyz/pull/733/commits/ddef7ec5acb3c516afce2a54bd13c8af08132be3